modules/collect/src/main/java/com/opengamma/strata/collect/io/CsvFile.java

```diff
@@ -88,29 +89,135 @@ public static CsvFile of(CharSource source, boolean headerRow) {
+    return new CsvFile(headers, buildSearchHeaders(headers), ImmutableList.copyOf(parsedCsv));
   }
 
-  //------------------------------------------------------------------------
   // parses the CSV file format
-  private static ArrayList<ImmutableList<String>> parse(ImmutableList<String> lines, char separator) {
-    return lines.stream()
-        .flatMap(line -> parseLine(line, separator))
-        .collect(toCollection(ArrayList::new));
+  private static ArrayList<ImmutableList<String>> parseAll(ImmutableList<String> lines, char separator) {
+    ArrayList<ImmutableList<String>> parsedLines = new ArrayList<>();
+    for (String line : lines) {
+      ImmutableList<String> parsed = parseLine(line, separator);
+      if (!parsed.isEmpty()) {
+        parsedLines.add(parsed);
+      }
+    }
+    return parsedLines;
   }
 
-  // return Stream rather than Optional as works better with flatMap
-  private static Stream<ImmutableList<String>> parseLine(String line, char separator) {
+  //------------------------------------------------------------------------
+  /**
+   * Obtains an instance from a list of headers and rows.
+   * <p>
+   * The headers may be an empty list.
+   * All the rows must contain a list of the same size, matching the header if present.
+   * 
+   * @param headers  the headers, empty if no headers
+   * @param rows  the data rows
+   * @return the CSV file
+   * @throws IllegalArgumentException if the rows do not match the headers
+   */
+  public static CsvFile of(List<String> headers, List<? extends List<String>> rows) {
+    ArgChecker.notNull(headers, "headers");
+    ArgChecker.notNull(rows, "rows");
+    int size = (headers.size() == 0 && rows.size() > 0 ? rows.get(0).size() : headers.size());
+    if (rows.stream().filter(row -> row.size() != size).findAny().isPresent()) {
+      throw new IllegalArgumentException("Invalid data rows, each row must have same columns as header row");
+    }
+    ImmutableList<String> copiedHeaders = ImmutableList.copyOf(headers);
+    ImmutableList<ImmutableList<String>> copiedRows = rows.stream()
+        .map(row -> ImmutableList.copyOf(row))
+        .collect(toImmutableList());
+    return new CsvFile(copiedHeaders, buildSearchHeaders(copiedHeaders), copiedRows);
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Parses the specified source as a CSV file without loading the whole file into memory.
+   * <p>
+   * The specified consumer is invoked for each row.
+   * 
+   * @param source  the CSV file resource
+   * @param headerRow  whether the source has a header row, an empty source may omit the header
+   * @param consumer  the row consumer
+   * @throws UncheckedIOException if an IO exception occurs
+   * @throws IllegalArgumentException if the file cannot be parsed
+   */
+  public static void parse(CharSource source, boolean headerRow, Consumer<CsvRow> consumer) {
```

Comment:
Designing the API to handle closing of the resource limits your options; you are bound to end up with an API which requires a stateful accumulator. I would prefer an API which leaves resource management to the user (which is simple) but allows an easier implementation for more complicated file formats.

There are parallels with XML parsing in Java. The original XML API was SAX which was conceptually similar to this API. The user provided a listener to the XML parser and the listener was necessarily stateful. SAX was painful to use and was superseded by the much easier to use StAX which has a pull-based API, much like an iterator.

We use a third-party CSV parser in one of our client projects which has complex CSV-based file formats. This API isn't sufficient for that use case and wouldn't enable us to remove the third-party parser. Migrating those file loaders to be stateful implementations of `Consumer` would make the already complicated code even more complex.